File: environments/k8s.go
package environments

type K8sAuthenticationConfig struct {
	Context *string `mapstructure:"context"`
	Config  *string `mapstructure:"config"`
}

func (k8s *K8sAuthenticationConfig) EnvVars() map[string]string {
	envVars := map[string]string{}

	if k8s.Context != nil {
		envVars["HELM_KUBECONTEXT"] = *k8s.Context
		envVars["KUBECTX"] = *k8s.Context
	}

	if k8s.Config != nil {
		envVars["KUBECONFIG"] = *k8s.Config
		envVars["KUBE_CONFIG_PATH"] = *k8s.Config
	}

	return envVars
}

func (dest *K8sAuthenticationConfig) Merge(src *K8sAuthenticationConfig) {
	if src.Context != nil {
		dest.Context = src.Context
	}
	if src.Config != nil {
		dest.Config = src.Config
	}
}

File: src/environments/k8s.ts
import { EnvironmentConfig } from './index'

export class K8sAuthenticationConfig implements EnvironmentConfig<K8sAuthenticationConfig> {
  context: string | undefined
  kubeconfig: string | undefined

  constructor(config: {
    context?: string, kubeconfig?: string
  }) {
    this.context = config.context
    this.kubeconfig = config.kubeconfig
  }

  env(): { [key: string]: string } {
    var env: { [key: string]: string } = {}
    if (this.context !== undefined) {
      env["HELM_KUBECONTEXT"] = this.context
      env["KUBECTX"] = this.context
    }

    if (this.kubeconfig !== undefined) {
      env["KUBECONFIG"] = this.kubeconfig
      env["KUBE_CONFIG_PATH"] = this.kubeconfig
    }

    return env
  }

  merge(...envs: K8sAuthenticationConfig[]): void {
    for (let env of envs) {
      if (env.context !== undefined) {
        this.context = env.context
      }

      if (env.kubeconfig !== undefined) {
        this.kubeconfig = env.kubeconfig
      }
    }
  }
}
